nano/node/bootstrap_ascending/database_scan.cpp

```diff
@@ -8,121 +8,158 @@
 {
+	auto composite = std::make_unique<container_info_composite> (name);
+	composite->add_component (std::make_unique<container_info_leaf> (container_info{ "accounts_iterator", accounts_iterator.completed, 0 }));
+	composite->add_component (std::make_unique<container_info_leaf> (container_info{ "pending_iterator", pending_iterator.completed, 0 }));
+	return composite;
 }
 
-nano::account nano::bootstrap_ascending::buffered_iterator::operator* () const
+/*
+ * account_database_iterator
+ */
+
+nano::bootstrap_ascending::account_database_iterator::account_database_iterator (nano::ledger & ledger_a) :
+	ledger{ ledger_a }
 {
-	return !buffer.empty () ? buffer.front () : nano::account{ 0 };
 }
 
-nano::account nano::bootstrap_ascending::buffered_iterator::next (std::function<bool (nano::account const &)> const & filter)
+std::deque<nano::account> nano::bootstrap_ascending::account_database_iterator::next_batch (nano::store::transaction & transaction, size_t batch_size)
 {
-	if (buffer.empty ())
+	std::deque<nano::account> result;
+
+	auto it = ledger.store.account.begin (transaction, next);
+	auto const end = ledger.store.account.end ();
+
+	for (size_t count = 0; it != end && count < batch_size; ++it, ++count)
 	{
-		fill ();
+		auto const & account = it->first;
+		result.push_back (account);
+		next = account.number () + 1;
 	}
 
-	while (!buffer.empty ())
+	if (it == end)
 	{
-		auto result = buffer.front ();
-		buffer.pop_front ();
-
-		if (filter (result))
-		{
-			return result;
-		}
+		// Reset for the next ledger iteration
+		next = { 0 };
+		++completed;
 	}
 
-	return { 0 };
+	return result;
 }
 
-bool nano::bootstrap_ascending::buffered_iterator::warmup () const
+bool nano::bootstrap_ascending::account_database_iterator::warmed_up () const
 {
-	return warmup_m;
+	return completed > 0;
 }
 
-void nano::bootstrap_ascending::buffered_iterator::fill ()
+/*
+ * pending_database_iterator
+ */
+
+nano::bootstrap_ascending::pending_database_iterator::pending_database_iterator (nano::ledger & ledger_a) :
+	ledger{ ledger_a }
 {
-	debug_assert (buffer.empty ());
+}
 
-	// Fill half from accounts table and half from pending table
-	auto transaction = ledger.tx_begin_read ();
+std::deque<nano::account> nano::bootstrap_ascending::pending_database_iterator::next_batch (nano::store::transaction & transaction, size_t batch_size)
+{
+	std::deque<nano::account> result;
 
-	for (int n = 0; n < size / 2; ++n)
-	{
-		accounts_iterator.next (transaction);
-		if (!(*accounts_iterator).is_zero ())
-		{
-			buffer.push_back (*accounts_iterator);
-		}
-	}
+	auto it = ledger.store.pending.begin (transaction, next);
+	auto const end = ledger.store.pending.end ();
 
-	for (int n = 0; n < size / 2; ++n)
-	{
-		pending_iterator.next (transaction);
-		if (!(*pending_iterator).is_zero ())
+	// TODO: This pending iteration heuristic should be encapsulated in a pending_iterator class and reused across other components
```

Comment:
Is this heuristic used to skip accounts with a large number of pending entries? A comment on why using sequential scanning alone is insufficient would help.